internal/connection_manager.go

```diff
@@ -63,36 +65,35 @@ func (cm *connectionManager) addListener(listener connectionListener) {
 	}
 }
 
-func (cm *connectionManager) getActiveConnection(address core.Address) *Connection {
-	if address == nil {
-		return nil
-	}
-	cm.lock.RLock()
-	defer cm.lock.RUnlock()
-	if conn, found := cm.connections[address.Host()+":"+strconv.Itoa(address.Port())]; found {
-		return conn
-	}
-	return nil
+func (cm *connectionManagerImpl) getActiveConnection(address core.Address) *Connection {
+	return cm.getConnection(address, false)
 }
 
-func (cm *connectionManager) getActiveConnections() map[string]*Connection {
+func (cm *connectionManagerImpl) ConnectionCount() int32 {
+	cm.connectionsMutex.RLock()
+	defer cm.connectionsMutex.RUnlock()
+	return int32(len(cm.connections))
+}
+
+func (cm *connectionManagerImpl) getActiveConnections() map[string]*Connection {
 	connections := make(map[string]*Connection)
-	cm.lock.RLock()
-	defer cm.lock.RUnlock()
+	cm.connectionsMutex.RLock()
+	defer cm.connectionsMutex.RUnlock()
 	for k, v := range cm.connections {
 		connections[k] = v
```

Comment:
Why don't we use write lock here?

Reply:
It looks confusing I guess.
`connections` is a local variable. I am only reading from cm.connections. Not writing to it. So a read lock is enough here.